dm/syncer/safe_mode.go

```diff
@@ -43,35 +56,68 @@ func (s *Syncer) enableSafeModeInitializationPhase(tctx *tcontext.Context) {
 		s.safeMode.Add(tctx, 1) // add 1 but has no corresponding -1, so keeps enabled
 		s.tctx.L().Info("enable safe-mode by config")
 	}
-	if s.checkpoint.SafeModeExitPoint() != nil {
-		//nolint:errcheck
-		s.safeMode.Add(tctx, 1) // enable and will revert after pass SafeModeExitLoc
-		s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *s.checkpoint.SafeModeExitPoint()))
+	exitPoint := s.checkpoint.SafeModeExitPoint()
+	failpoint.Inject("SafeModeDurationIsZero", func() {
+		exitPoint = nil
+	})
+	if exitPoint != nil {
+		if binlog.CompareLocation(*exitPoint, s.getInitExecutedLoc(), s.cfg.EnableGTID) > 0 {
```

Comment:
seems initExecutedLoc is only recorded once when Syncer is started, it will not be updated when subtask pause/resume.

Reply:
I don't think so. `Syncer.Process()` is also called by `Syncer.Resume()`

Reply:
`initExecutedLoc` will be set only when it's `nil`, so it will not be written again when pause/resume

Reply:
You are right! I added `beginLocation` in `Syncer`, and use it to compare with `exitPoint`. `beginLocation` will init whatever start/resume.